### DISCUSSION TITLE
Transactional Parallel Consumer stuck while rebalancing

### DISCUSSION THREAD

**Discussion by mikhelef** [confluentinc/parallel-consumer#542] [Q&A]
Hi,
When starting multiple instances of parallelconsumer and having an input topic full of records, the first instance starts consuming and processsing records and then get stuck while rebalancing (when the other instances join the group) because it cannot produce the remaining records to the destination topic as it cannot get the produce lock (maybe because it is locked by the flush when partition is revoked).
After max.poll.interval.ms (6mn) the consumer is considered dead and exit the consumer group.

the used configuration is:
- PC version: 0.5.2.4
- commitMode = PERIODIC_TRANSACTIONAL_PRODUCER
- producelock = 2mn
- max.poll.records = 48
- max.poll.interval.ms = 6mn
- PC concurrency per instance (threads) = 16
- Number of partitions = 4
- Number of intances / kafka consumers = 4
- max.request.size: 30000000
- batch.size: 250000
- linger.ms: 65

I have tried increasing producelock timeout to 4 minutes but got the same result, one or multiple consumers stuck at rebalancing then got excluded from the consumer group.

Has anyone experienced this ? 

Thank you.

**Comment by astubbs** [comment]
@mikhelef - your report is the only field description of this failure mode, and its shape matches what the maintained fork tracks as astubbs/parallel-consumer#44 (this tracker's #803): transactional mode, a second instance joins, the first stops consuming and is evicted after `max.poll.interval.ms`.

The mechanism, which explains why raising the produce-lock timeout changed nothing: the revocation callback runs on the poll thread, and on 0.5.x it spun there waiting for the in-flight transaction commit to finish. While it waited, your workers were blocked behind the same lock - that is the "cannot get the produce lock" you saw - and the poll thread was not polling, so the coordinator evicted the member once `max.poll.interval.ms` ran out. The produce-lock timeout is not the wait that evicts you, so two minutes or four made no difference; the duration was never the variable.

Where it stands on [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0): the revocation commit is handed to the control thread instead of spinning on the poll thread, and the callback's wait is bounded by `commitLockAcquisitionTimeout` rather than unbounded (astubbs/parallel-consumer#466). Measured on a 20-second in-flight commit, the old spin held the poll thread for 79 seconds because the control thread kept re-taking the lock as fast as it dropped it; it now holds it for the one commit. What is not yet fixed, and the release notes name it as a known limitation: that bound defaults to five minutes, which can still exceed `max.poll.interval.ms`, so a long enough commit during a rebalance still evicts the member. Bounding it to something under the poll interval is astubbs/parallel-consumer#408, open.

If you are still running this shape, the release is worth a try on your four-instance setup; a report of what you see on it would go on astubbs/parallel-consumer#44.

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
